traffic-advice.bs

```diff
@@ -0,0 +1,265 @@
+
+1.  Return a new [=URL=] as follows:
+
+    :  [=url/scheme=]
+    :: |origin|'s [=origin/scheme=]
+    :  [=url/host=]
+    :: |origin|'s [=origin/host=]
+    :  [=url/port=]
+    :: |origin|'s [=origin/port=]
+    :  [=url/path=]
+    :: « `".well-known"`, `"traffic-advice"` »
+    
+</section>
+
+<section algorithm="fetch-traffic-advice">
+
+To <dfn>fetch traffic advice</dfn> for [=origin=] |origin|, [=agent identity=] |identity| and algorithm |whenComplete| accepting a [=traffic advice result=]:
+
+1.  Let |url| be the result of [=generating a traffic advice URL=] for |origin|.
+    If it results in failure, then return failure.
+
+1.  Let |request| be a [=request=] as follows:
+
+    :  [=request/method=]
+    :: `` `GET` ``
+    :  [=request/URL=]
+    :: |url|
+    :  [=request/client=]
+    :: null
+    :  [=request/credentials mode=]
+    :: `"omit"`
+    :  [=request/redirect mode=]
+    :: `"manual"`
+        <div class="note">This means that a [=redirect status=] will not lead to another origin being contacted.</div>
+
+1.  Let |processResponse| be the following steps, given [=response=] |response|:
+
+    1.  If |response|'s [=response/type=] is `"error"`, then terminate the fetch, run |whenComplete| with `"unreachable"`, and return.
+
+    1.  If |response|'s [=response/type=] is `"opaqueredirect"`, then terminate the fetch, run |whenComplete| with null, and return.
+
+    1.  [=Assert=]: |response|'s [=response/type=] is `"basic"`.
+
+    1.  If |response|'s [=response/status=] is 429 (Too Many Requests; see [[RFC6585]]) or 503 (Service Unavailable; see [[HTTP-SEMANTICS]]), then terminate the fetch, run |whenComplete| with `"unreachable"`, and return.
+        <div class="note">If present, the [[HTTP-SEMANTICS]] `Retry-After` response header could be used as a hint about when to next retry.</div>
+
+    1.  If |response|'s [=response/status=] is not an [=ok status=], then terminate the fetch, run |whenComplete| with null and return.
+
+    1.  If |response|'s [=response/status=] is a [=null body status=], then terminate the fetch, run |whenComplete| with null and return.
+
+    1.  Let |mimeType| be the result of [=header list/extracting a MIME type=] from |response|'s [=response/header list=].
+
+    1.  If |mimeType| is failure or its [=MIME type/essence=] is not `"application/trafficadvice+json"`, then terminate the fetch, run |whenComplete| with null and return.
+
+1.  Let |processResponseEndOfBody| be the following steps, given [=response=] |response| and null, failure or [=byte sequence=] |body|:
+
+    1.  If |body| is not a [=byte sequence=], then run |whenComplete| with null and return.
+
+    1.  Let |string| be the result of [=UTF-8 decoding=] |body|.
+
+    1.  Let |parseResult| be the result of [=parsing traffic advice=] from |string| given |identity|.
+
+    1.  Run |whenComplete| with |parseResult|.
+
+1.  [=Fetch=] |request| with |processResponse| and |processResponseEndOfBody|.
```

Comment:
The style per https://infra.spec.whatwg.org/#algorithm-params is something like

```html
[=Fetch=] |request| with <i>[=fetch/processResponse=]</i> set to |processResponse| and <i>[=fetch/processResponseEndOfBody=]</i> set to |processResponseEndOfBody|.
```

Reply:
Done.